lib/Ethereum/RPC/Contract.pm

```diff
@@ -15,57 +15,80 @@ our $VERSION = '0.001';
 use Moo;
 use JSON::MaybeXS;
 use Math::BigInt;
+use Scalar::Util qw(looks_like_number);
+use List::Util qw(first);
 
 use Ethereum::RPC::Client;
 use Ethereum::RPC::Contract::ContractResponse;
 use Ethereum::RPC::Contract::ContractTransaction;
 use Ethereum::RPC::Contract::Helper::UnitConversion;
 
-has contract_address => ( is => 'rw' );
-has contract_abi     => ( is => 'ro', required => 1 );
-has rpc_client       => ( is => 'ro', lazy => 1 );
+has contract_address => (is => 'rw');
+has contract_abi => (
+    is       => 'ro',
+    required => 1
+);
+has rpc_client => (
+    is => 'lazy',
+);
 
 sub _build_rpc_client {
     return Ethereum::RPC::Client->new;
 }
 
-has from             => ( is => 'rw', lazy => 1 );
+has from => (
+    is   => 'rw',
+    lazy => 1
+);
 
 sub _build_from {
     return shift->rpc_client->eth_coinbase();
 }
 
-has gas_price        => ( is => 'rw', lazy => 1 );
+has gas_price => (
+    is   => 'rw',
+    lazy => 1
+);
 
 sub _build_gas_price {
     return shift->rpc_client->eth_gasPrice();
 }
 
-has gas              => ( is => 'rw' );
+has gas => (is => 'rw');
 
-has contract_decoded => ( is => 'ro', default => sub{{}} );
+has contract_decoded => (
+    is      => 'rw',
+    default => sub{{}},
+);
 
 =head2 BUILD
 
-Constructor: Here we get all functions from the passed ABI and bring it to contract class subs.
+Constructor: Here we get all functions and events from the given ABI and set
+it to the contract class.
 
-Parameters:
-    contract_address    ( Optional - only if the contract already exists ),
-    contract_abi        ( Required - https://solidity.readthedocs.io/en/develop/abi-spec.html ),
-    rpc_client          ( Optional - Ethereum::RPC::Client(https://github.com/binary-com/perl-Ethereum-RPC-Client) - if not given, new instance will be created );
-    from                ( Optional - Address )
-    gas                 ( Optional - Integer gas )
-    gas_price           ( Optional - Integer gasPrice )
+=over 4
 
-Return:
-    New contract instance
+=item contract_address => string (optional)
+
+=item contract_abi => string (required, https://solidity.readthedocs.io/en/develop/abi-spec.html)
+
+=item rpc_client => L<Ethereum::RPC::Client> (optional, default: L<Ethereum::RPC::Client>)
+
+=item from => string (optional)
+
+=item gas => numeric (optional)
+
+=item gas_price => numeric (optional)
+
+=back
 
 =cut
 
 sub BUILD {
     my ($self) = @_;
 
-    my @decoded_json = @{decode_json($self->contract_abi)};
+    my $json = $self->contract_abi ? decode_json($self->contract_abi) : undef;
+    my @decoded_json = ref $json eq 'ARRAY' ? @$json : [];
```

Comment:
Probably `: ()` instead of `: []` as `@decoded_json` is an actual array.

Might also be worth rephrasing as thus

```perl
my @decoded_json = decode_json( $self->contract_abi // "[]" );
```